File: Search/Exponential_Search/exponential_search.py
def binary_search(arr, low, high, x):
    while low <= high:
        mid = (low + high) // 2
        if arr[mid] == x: return mid
        if arr[mid] < x: low = mid + 1
        else: high = mid - 1
    return -1

def exponential_search(arr, x):
    if arr[0] == x: return 0
    
    n = len(arr)
    i = 1
    while i < n and arr[i] <= x:
        i = i * 2
        
    return binary_search(arr, i // 2, min(i, n - 1), x)

if __name__ == "__main__":
    data = [2, 3, 4, 10, 40, 50, 70, 80, 100]
    target = 70
    result = exponential_search(data, target)
    print(f"Element found at index: {result}" if result != -1 else "Not found")

File: Search/Exponential_Search/exponential_search.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int binarySearch(const std::vector<int>& arr, int l, int r, int x) {
    while (l <= r) {
        int m = l + (r - l) / 2;
        if (arr[m] == x) return m;
        if (arr[m] < x) l = m + 1;
        else r = m - 1;
    }
    return -1;
}

int exponentialSearch(const std::vector<int>& arr, int x) {
    if (arr[0] == x) return 0;
    int n = arr.size();
    int i = 1;
    while (i < n && arr[i] <= x) i = i * 2;
    return binarySearch(arr, i / 2, std::min(i, n - 1), x);
}

int main() {
    std::vector<int> data = {2, 3, 4, 10, 40, 50, 70, 80, 100};
    int x = 70;
    int result = exponentialSearch(data, x);
    if (result != -1) std::cout << "Found at index " << result;
    else std::cout << "Not found";
    return 0;
}
